src/peripheral.rs

```diff
@@ -0,0 +1,45 @@
+use crate::descriptors::gatt_service::GattService;
+use crate::gap_advertiser::GapAdvertiser;
+use crate::gatt_connection::GattConnection;
+use crate::gatt_server_cb::GattServerCallback;
+use core::fmt::Debug;
+
+/// Trait defining the capabilities of a BLE peripheral device, typically found on embedded
+/// devices.  For more information, see:
+/// https://embeddedcentric.com/lesson-2-ble-profiles-services-characteristics-device-roles-and-network-topology/
+///
+/// Note that this trait is meant to be a low-level mapping to common BLE libraries and not
+/// ergonomic and safe for Rust users.
+///
+/// The implementation should be embedded friendly as it is common for BLE peripherals to run on
+/// embedded platforms.
+pub trait Peripheral {
+  type SystemError: Debug;
+
+  type Handle;
+  type Advertiser: GapAdvertiser + Debug + Clone;
+  type Connection: GattConnection + Debug;
+
+  /// Sets the device name that can be read from built-in GATT characteristics after connection,
+  /// and a separate, shorter version can be sent through the advertisement packet.  There may
+  /// be different limits imposed by each platform.
+  fn set_name(&mut self, name: &str) -> Result<(), Self::SystemError>;
+
+  /// Sets the device appearance (generally this refers to how the device should be visually
+  /// or functionally presented to a user).  This data is presented upon connection using
+  /// mandatory built-in GATT characteristics.
+  fn set_appearance(&mut self, appearance: u16) -> Result<(), Self::SystemError>;
+
+  /// Configure a new GATT server on this peripheral.  Note that only one GATT
+  /// server may be configured for each peripheral but this server may contain multiple distinct
+  /// GATT services.  Connections will not be accepted until [GapAdvertiser::request_start] is
+  /// called and the advertisement explicitly enables connectability.
+  ///
+  /// The entire GATT server application is expected to be driven through the event callback.
+  /// Dropping the resulting handle will shutdown the server.
+  fn configure_gatt_server(
+    self,
+    services: &[GattService],
+    callback: impl GattServerCallback<Self> + 'static,
```

Comment:
* `'static` is too constraining, ideally should be `'a` and `Handle` should be `Handle<'a>`
* Needs to be `Send` as discussed above


Also, this whole `Peripheral` thing looks a bit like a builder, yet it does not have enough meat to be a builder. Why not a single factory method - `create` - that just takes name, appearance, services and a callback? Ideally, `&self` so that `Peripheral` a.k.a. the "driver" can be used as a factory for other stuff too (BT classic servers also come to mind)?

Reply:
Regarding names:
* How about `Handle` renamed to `GattServer`
* How about `Peripheral` renamed to `Driver`
?

Reply:
Btw and if that's not clear, I'm assuming that `Handle` / `GattServer` would be roughly analogous to - and implemented on - [this](https://github.com/esp-rs/esp-idf-svc/blob/master/src/bt/ble/gatt/server.rs#L347).

As for `Perihperal` (that I used to call `Driver` in my comments, but yes, ok, good reasoning to stay as it is) - it would be roughly analogous and implement on [that](https://github.com/esp-rs/esp-idf-svc/blob/master/src/bt.rs#L388).

As you can imagine, you can create multiple - eerrrr - "servers" on a single `Peripheral` (= `BtDriver` in `esp-idf-svc`-speak).

Hence `&self` and not `&mut self` and probably a runtime error if you try to create two GATT servers. :)

Reply:
A lot to unpack here, and great comments overall for sure.  A few quick points and then I'll try to internalize the rest and rethink if there's a more clear expression of this code:

1. For sure about `'static` and `Send`, that felt like a total hack to just get past some nasty compiler errors I was seeing.  I'll rethink.
2. You actually can't have multiple GATT servers on a single Peripheral -- there's a 1:1 mapping between all of a  physical Bluetooth radio, a peripheral GAP role, and a GATT server role.  The reason these are separate concepts is because you can mix-and-match them: you can have a physcial radio that implements a peripheral role but a GATT client role, or a physical radio that's a central role and a GATT client.  Once you pick though, you're stuck and you can't mix and match after the fact.  Note that these traits conflate that a GAP peripheral is always a GATT server because, in practice, there is absolutely no reason to use a GAP peripheral as a GATT client.  The basic concept is absurd enough that I've never seen a single library API that allows you to swap mix-and-match the two roles.  I think it's just because of the obtuse nature of the layered Bluetooth spec.
3. The reason we call this Peripheral and Handle are to show that the GattServer role is encapsulated by the Peripheral role.  The hierarchy is radio > gap role (central/peripheral) > gatt role (client/server).  Our API is letting you set up the GAP role and implying the GATT role.  I did conflate these two as GattServerCallback though because it is entirely too obtuse and awkward to split the GAP callbacks from the GATT ones from Rust's point of view because then you've gotta implement a bunch of extra message passing since they're so intertwined they virtually never can be implemented independent of each other.